bin/metrics-logstash-node.rb

```diff
@@ -113,31 +113,63 @@ def run # rubocop:disable Metrics/AbcSize
-      item['events'] = {} unless item.key?('events')
-      metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.events.in"] = item['events']['in'].to_i || 0
-      metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.events.out"] = item['events']['out'].to_i || 0
-      metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.events.duration_in_millis"] = item['events']['duration_in_millis'].to_i || 0
-      metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.matches"] = item['matches'].to_i if item.key?('matches')
-    end
-
-    node['pipeline']['plugins']['outputs'].each do |item|
-      item['events'] = {} unless item.key?('events')
-      metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.in"] = item['events']['in'].to_i || 0
-      metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.out"] = item['events']['out'].to_i || 0
-      metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.duration_in_millis"] = item['events']['duration_in_millis'].to_i || 0
+    # logstash < 6.0
+    if node.key?('pipeline')
+      node['pipeline']['events'].each do |key, value|
+        metrics["pipeline.events.#{key}"] = value
+      end
+
+      node['pipeline']['plugins']['inputs'].each do |item|
+        item['events'] = {} unless item.key?('events')
+        metrics["pipeline.plugins.inputs.#{item['name']}.#{item['id']}.events.in"] = item['events']['in'].to_i || 0
+        metrics["pipeline.plugins.inputs.#{item['name']}.#{item['id']}.events.out"] = item['events']['out'].to_i || 0
+        metrics["pipeline.plugins.inputs.#{item['name']}.#{item['id']}.events.queue_push_duration_in_millis"] = \
+          item['events']['queue_push_duration_in_millis'].to_i || 0
+      end
+
+      node['pipeline']['plugins']['filters'].each do |item|
+        metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.events.out"] = item['events']['out'].to_i || 0
+        metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.events.duration_in_millis"] = item['events']['duration_in_millis'].to_i || 0
+        metrics["pipeline.plugins.filters.#{item['name']}.#{item['id']}.matches"] = item['matches'].to_i if item.key?('matches')
+      end
+
+      node['pipeline']['plugins']['outputs'].each do |item|
+        item['events'] = {} unless item.key?('events')
+        metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.in"] = item['events']['in'].to_i || 0
+        metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.out"] = item['events']['out'].to_i || 0
+        metrics["pipeline.plugins.outputs.#{item['name']}.#{item['id']}.events.duration_in_millis"] = item['events']['duration_in_millis'].to_i || 0
+      end
+    # logstash >= 6.0
+    elsif node.key?('pipelines')
+      node['pipelines'].each_key do |pipeline|
+        unless node['pipelines'][pipeline]['events'].nil?
```

Comment:
Can't assume that `node['pipelines'][pipeline]['events']` will always be a valid object, I've observed this to occasionally be a string value of `null`

Is this the correct way to handle this?

Reply:
if if it falls into these three categories:
- string (valid)
- string 'null' (invalid)
- nil

I think you need do some extra stuff...

```ruby
if node['pipelines'][pipeline]['events'].nil? || node['pipelines'][pipeline]['events'] == 'null'
  # skip me
# else
  # do foo
end

``` 

That being said there are far too many nested conditionals here so the code is kinda unwieldy and will be slow, I think we need to break some of this out to additional functions.